research/gaia-x/pegasus/pegasus/src/operator/concise/limit.rs

```diff
@@ -53,3 +60,158 @@ impl<D: Data> Limit<D> for Stream<D> {
         })
     }
 }
+
+impl<D: Data + Ord> SortLimit<D> for Stream<D> {
+    fn sort_limit(self, size: u32) -> Result<Stream<D>, BuildJobError> {
+        if size == 0 {
+            return BuildJobError::unsupported("sort_limit n cannot equal to zero");
+        }
+        self.sort_limit_by(size, |x, y| x.cmp(y))
+    }
+}
+
+impl<D: Data> SortLimitBy<D> for Stream<D> {
+    fn sort_limit_by<F>(self, size: u32, cmp: F) -> Result<Stream<D>, BuildJobError>
+    where
+        F: Fn(&D, &D) -> Ordering + Send + 'static,
+    {
+        if size == 0 {
+            return BuildJobError::unsupported("sort_limit n cannot equal to zero");
+        }
+        let share_cmp = ShadeCmp { cmp: Rc::new(cmp) };
+        let cmp_clone = ShadeCmp { cmp: share_cmp.cmp.clone() };
+        let local_sort = sort_limit_by_partition(self, size, cmp_clone)?;
+        sort_limit_by_partition(local_sort.aggregate(), size, share_cmp)
+    }
+}
+
+#[inline]
+fn sort_limit_by_partition<D: Data, F>(
+    stream: Stream<D>, size: u32, share_cmp: ShadeCmp<F>,
+) -> Result<Stream<D>, BuildJobError>
+where
+    F: Fn(&D, &D) -> Ordering + Send + 'static,
+{
+    stream.unary("sort_limit_by_partition", |info| {
+        let mut table = TidyTagMap::new(info.scope_level);
+        move |input, output| {
+            input.for_each_batch(|dataset| {
+                if !dataset.is_empty() {
+                    let custom_heap = table.get_mut_or_else(&dataset.tag, || {
+                        FixedSizeHeap::new(size as usize, share_cmp.cmp.clone())
+                    });
+                    for d in dataset.drain() {
+                        custom_heap.add(d)?;
+                    }
+                }
+                if dataset.is_last() {
+                    let mut session = output.new_session(&dataset.tag)?;
+                    if let Some(custom_heap) = table.remove(&dataset.tag) {
+                        session.give_iterator(custom_heap.into_iter())?
+                    }
+                }
+                Ok(())
+            })
+        }
+    })
+}
+
+struct FixedSizeHeap<D, C: Fn(&D, &D) -> Ordering + Send + 'static> {
+    pub limit: usize,
+    cmp: Rc<C>,
+    pub heap: BinaryHeap<Item<D, C>>,
+}
+
+impl<D, C: Fn(&D, &D) -> Ordering + Send + 'static> IntoIterator for FixedSizeHeap<D, C> {
+    type Item = D;
+    type IntoIter = std::vec::IntoIter<D>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let mut vec = Vec::with_capacity(self.heap.len());
+        for item in self.heap.into_sorted_vec() {
+            vec.push(item.inner);
+        }
+        vec.into_iter()
+    }
+}
+
+impl<D, C: Fn(&D, &D) -> Ordering + Send + 'static> FixedSizeHeap<D, C> {
```

Comment:
Better provide one constructor with `D: Ord` and the default comparater of x.cmp(y) if both are `Ord`. This makes it easier to use.